README.md

```diff
@@ -175,10 +176,16 @@ Note: `document.cookie` does not get cleaned up automatically. You'll have to cl
 
 Next.js `Link` component invokes `window.scrollTo` on click which is not implemented in JSDOM environment. In order to fix the error you should [set up your test environment](#set-up-your-test-environment) or provide [your own `window.scrollTo` mock](https://qiita.com/akameco/items/0edfdae02507204b24c8).
 
+### Warning: Text content did not match. Server: "x" Client: "y"" error
+
+This warning can appear during client side hydration of the page and can be due to a bug in your code or due to to the fact that `next-page-tester` has to isolate modules in `server` and `client` environment to avoid "environment leaking". In second case, modules that needs to use same instance in `server` and `client` environment (e.g. `React.Context`, or `css-in-js` libraries) can be marked as such by being passed in the `nonIsolatedModules` option of the `getPage` function; [styletron-react example](src/__tests__/third-party/styletron-react/styletron-react.test.ts#L13)
```

Comment:
It makes sense. I know in actual SSR application things work differently, but if I understand our current approach correctly, we need to preserve module identity for modules that are using "React" constucts to avoid issues.

Reply:
The fact that we use the same module identity between server and client renders it's actually a quite nasty workaround, which I expect leading to some bug sooner or later.

What we'd need (and we don't have yet) is preserving imported module identity for a whole rendering cycle in a given environment: 1 for server render, a new one for client render, and a new one for every subsequent data fetching call.

Since we don't have parallel separated environment we switch between server and client: client is currently the same natural scope where tests run, while we try to enclose server operation in `executeAsIfOnServer` and `requireAsIfOnServer ` scopes.

Since each call to `requireAsIfOnServer` generates fresh modules, it's here where we break code which relies on module identity. Therefore we whitelist modules which identity we want to preserve throughout the whole session, with the side effect of letting their identity leaking into `client` execution (which should have its own modules instances).

Last point: `Text content did not match` could be due to a bug in userland code but it just signal that server and client render output are different.

Reply:
Quick though that passed my mind and maybe an option to explore one day.

What if we did SSR in a separate process (we would spawn a new process without JSDOM environment) and just do the server render there and we would get the "html" from it. Everything that happens in the main process is "client" environment. 

If this works conceptually, we would get rid of all these problems.

Reply:
I also wondered whether a similar approach could be feasible. Since I've never manually spawned Node.js processes I've never tried out. I wonder whether it could have downsides in terms of handling errors thrown in spawn processes. I'd love to explore this direction.

Also, there is another simpler option which which I still haven't investigated (but I suspect might fall short of flexibility): which is keeping our rendering phases linear and avoid mixing server and client logic. We emulate sever render first and client render afterwards.